sql/core/src/test/scala/org/apache/spark/sql/DataFrameSetOperationsSuite.scala

```diff
@@ -808,6 +813,53 @@ class DataFrameSetOperationsSuite extends QueryTest with SharedSparkSession {
     // scalastyle:on
     checkAnswer(union, row1 :: row2 :: Nil)
   }
+
+  test("SPARK-34474: Remove unnecessary Union under Distinct") {
+    Seq(RemoveNoopUnion.ruleName, "").map { ruleName =>
+      withSQLConf(SQLConf.OPTIMIZER_EXCLUDED_RULES.key -> ruleName) {
+        val distinctUnionDF1 = testData.union(testData).distinct()
+        checkAnswer(distinctUnionDF1, testData.distinct())
+
+
+        val distinctUnionDF2 = testData.union(testData).dropDuplicates(Seq("key"))
+        checkAnswer(distinctUnionDF2, testData.dropDuplicates(Seq("key")))
+
+        val distinctUnionDF3 = sql(
+          """
+            |select key, value from testData
+            |union
+            |select key, value from testData
+            |""".stripMargin)
+        checkAnswer(distinctUnionDF3, testData.distinct())
+
+        val distinctUnionDF4 = sql(
+          """
+            |select distinct key, expr
+            |from
+            |(
+            |  select key, key + 1 as expr
+            |  from testData
+            |  union all
+            |  select key, key + 2 as expr
+            |  from testData
+            |)
```

Comment:
After checking the optimized logical plan of this query, I found the `collect` in `RemoveNoopUnion` relies on the fact that the query has one extra unnecessary `Project` on one child:

```
Distinct
+- Union false, false
   :- Project [key#3, (key#3 + 1) AS expr#143]
   :  +- SerializeFromObject [knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$TestData, true])).key AS key#3, staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$TestData, true])).value, true, false) AS value#4]
   :     +- ExternalRDD [obj#2]
   +- Project [key#3 AS key#145, expr#144 AS expr#146]. <<-- this one
      +- Project [key#3, (key#3 + 2) AS expr#144]
         +- SerializeFromObject [knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$TestData, true])).key AS key#3, staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$TestData, true])).value, true, false) AS value#4]
            +- ExternalRDD [obj#2]
```

So the rules works because there's one unnecessary `Project` existed before the rule. And later on this `Project` is removed by `CollapseProject`. If this `Project` does not exist, the correctness of the `RemoveNoopUnion` rule cannot be guaranteed.

I feel this is dangerous and may lead to wrong query plan later. But I am also fine that we guard this with a unit test here.

btw, because of the unnecessary `Project` in the middle, this query cannot be optimized by the rule:

```
select distinct key, expr
from
(
  select key, key as expr
  from testData
  union all
  select key, key as expr
  from testData
)
```

Reply:
You can verify it by excluding also `CollapseProject`. It doesn't affect the rule.

Reply:
@viirya - I am not saying without `CollapseProject` this rule is wrong. You put the rule before `CollapseProject`, so it is not affected by that anyway. I am saying the rule depends on the unnecessary `Project` implicitly, and rule can go wrong / sub-optimal. E.g., above query cannot be optimized by the rule.

Reply:
Manually put `CollapseProject` ahead of this rule and the project is removed, but I don't see it is affected too. Can you check it again?

Reply:
@viirya - I am not sure if we are in the same page now. This is a counter-example to produce wrong query plan if we have `CollapseProject` ahead of the rule - https://github.com/apache/spark/compare/80b03dd64e317...c21:remove-union .

Reply:
I posted the query plan with/without this rule for the query from the diff in the above link. I don't see the query plan is different. I'm not sure which query plan is wrong?

```
sql(
  """
     |select key, key + 2 as expr
     |from testData4
     |""".stripMargin)
```

```
== Physical Plan ==                                                                                                                                                                                                                    
*(1) Project [key#2641, (key#2641 + 2) AS expr#2645]                                                               
+- *(1) SerializeFromObject [knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$TestData4, true])).key AS key#2641]                                                                                             
   +- Scan[obj#2640]    
```

```
== Physical Plan ==
*(1) Project [key#2732, (key#2732 + 2) AS expr#2736]
+- *(1) SerializeFromObject [knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$TestData4, true])).key AS key#2732]
   +- Scan[obj#2731]
```

Reply:
@viirya - in https://github.com/apache/spark/compare/80b03dd64e317...c21:remove-union,  I am testing this query:

```
select distinct key, expr
from
(
  select key, key + 1 as expr
  from testData
  union all
  select key, key + 2 as expr
  from testData
)
```

Optimized logical plan with the rule:

```
== Optimized Logical Plan ==
Aggregate [key#14, expr#18], [key#14, expr#18]
+- Project [key#14, (key#14 + 1) AS expr#18]
   +- SerializeFromObject [knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$TestData4, true])).key AS key#14]
      +- ExternalRDD [obj#13]
```

Optimized logical plan without the rule:

```
== Optimized Logical Plan ==
Aggregate [key#2, expr#6], [key#2, expr#6]
+- Union false, false
   :- Project [key#2, (key#2 + 1) AS expr#6]
   :  +- SerializeFromObject [knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$TestData4, true])).key AS key#2]
   :     +- ExternalRDD [obj#1]
   +- Project [key#2, (key#2 + 2) AS expr#9]
      +- SerializeFromObject [knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$TestData4, true])).key AS key#2]
         +- ExternalRDD [obj#1]
```

Reply:
Oh, thanks. I think it is clear now. Let me fix it.